src/workers/pool.ts

```diff
@@ -0,0 +1,369 @@
+}
+
+export class ChildProcessWorkerOps<I> implements WorkerOps<I, ChildProcess> {
+  public constructor(private readonly main: string) {}
+
+  public start(): ChildProcess {
+    return fork(join(__dirname, 'child-process-worker.js'), [this.main])
+  }
+
+  public stop(worker: ChildProcess): void {
+    worker.kill()
+  }
+
+  public send(worker: ChildProcess, message: I): void {
+    worker.send(message)
+  }
+
+  public describe(worker: ChildProcess): string {
+    return `ChildProcess { pid: ${worker.pid} }`
+  }
+}
+
+export class WorkerThreadWorkerOps<I> implements WorkerOps<I, Worker> {
+  public constructor(private readonly main: string) {}
+
+  public start(): Worker {
+    return new Worker(join(__dirname, 'worker-thread-worker.js'), {
+      workerData: { __workerPath: this.main },
+    })
+  }
+
+  public stop(worker: Worker): void {
+    worker.terminate()
+  }
+
+  public send(
+    worker: Worker,
+    message: I,
+    transferList?: (ArrayBuffer | MessagePort)[]
+  ): void {
+    worker.postMessage(message, transferList)
+  }
+
+  public describe(worker: Worker): string {
+    return `Worker { threadId: ${worker.threadId} }`
+  }
+}
+
+export class InlineWorkerOps<I, O> implements WorkerOps<I, EventEmitter> {
+  private workerInstance: EventEmitter
+
+  public constructor(private readonly call: (input: I) => Promise<O>) {
+    this.workerInstance = new EventEmitter()
+  }
+
+  public start(): EventEmitter {
+    return this.workerInstance
+  }
+
+  public stop(worker: EventEmitter): void {
+    if (worker !== this.workerInstance) {
+      throw new Error('worker is not owned by this instance')
+    }
+
+    worker.removeAllListeners()
+  }
+
+  public async send(worker: EventEmitter, message: I): Promise<void> {
+    try {
+      const output = await this.call(message)
+      worker.emit('message', { output })
+    } catch (error) {
+      worker.emit('error', error)
+    }
+  }
+
+  public describe(): string {
+    return 'Worker { inline: true }'
+  }
+}
+
+export class WorkerPool<I, O, W extends EventEmitter = EventEmitter> {
+  private workers?: Set<W>
+  private claimedWorkers?: Set<W>
+  private outstandingClaims: Deferred<W>[] = []
+  private outstandingPerWorkerClaims = new Map<W, Deferred<W>[]>()
+
+  constructor(
+    private readonly workerOps: WorkerOps<I, W>,
+    private readonly size: number
+  ) {}
+
+  public start(): void {
+    if (this.workers) {
+      throw new Error('cannot start when already started')
+    }
+
+    const workers = new Set<W>()
+
+    for (let i = 0; i < this.size; i++) {
+      const worker = this.workerOps.start()
+      debug('started worker #%d: %o', i, this.workerOps.describe(worker))
+      workers.add(worker)
+    }
+
+    this.workers = workers
+    this.claimedWorkers = new Set()
+    debug('started %d worker(s)', this.workers.size)
+  }
+
+  public stop(): void {
+    const { workers } = this
+
+    if (!workers) {
+      return
+    }
+
+    delete this.workers
```

Comment:
should this delete happen after all the processing? maybe in a `finally` ?

Reply:
Unsure. I wanted to try to prevent any more calls from happening and I'm using the presence of `this.workers` as a way of saying "running or not". Maybe I should have an explicit state for that: stopped / starting / started / stopping / stopped.